server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsService.java

```diff
@@ -420,13 +407,67 @@ public void updateSettings(
+        boolean dryRun,
+        ActionListener<DataStream> listener
     ) {
-        updateSettingsTaskQueue.submitTask(
-            "updating settings on data stream",
-            new UpdateSettingsTask(projectId, dataStreamName, settingsOverrides, ackTimeout, listener),
-            masterNodeTimeout
+        if (dryRun) {
+            /*
+             * If this is a dry run, we'll do the settings validation and apply the changes to the data stream locally, but we won't run
+             * the task that actually updates the cluster state.
+             */
+            try {
+                DataStream updatedDataStream = createDataStreamForUpdatedDataStreamSettings(
+                    projectId,
+                    dataStreamName,
+                    settingsOverrides,
+                    clusterService.state()
+                );
+                listener.onResponse(updatedDataStream);
+            } catch (Exception e) {
+                listener.onFailure(e);
+            }
+        } else {
+            UpdateSettingsTask updateSettingsTask = new UpdateSettingsTask(
+                projectId,
+                dataStreamName,
+                settingsOverrides,
+                clusterService,
+                ackTimeout,
+                listener
+            );
+            updateSettingsTaskQueue.submitTask("updating settings on data stream", updateSettingsTask, masterNodeTimeout);
+        }
+    }
+
+    /*
+     * This method validates that the settings won't cause any validation problems with existing templates. If successful, a copy of the
+     * data stream is returned with the new settings applied.
+     */
+    private DataStream createDataStreamForUpdatedDataStreamSettings(
+        ProjectId projectId,
+        String dataStreamName,
+        Settings settingsOverrides,
+        ClusterState clusterState
+    ) throws Exception {
+        ProjectMetadata projectMetadata = clusterState.metadata().getProject(projectId);
+        Map<String, DataStream> dataStreamMap = projectMetadata.dataStreams();
+        DataStream dataStream = dataStreamMap.get(dataStreamName);
+        Settings existingSettings = dataStream.getSettings();
+
+        Template.Builder templateBuilder = Template.builder();
+        Settings.Builder mergedSettingsBuilder = Settings.builder().put(existingSettings).put(settingsOverrides);
+        Settings mergedSettings = mergedSettingsBuilder.build();
+
+        final ComposableIndexTemplate template = lookupTemplateForDataStream(dataStreamName, projectMetadata);
+        ComposableIndexTemplate mergedTemplate = template.mergeSettings(mergedSettings);
+        MetadataIndexTemplateService.validateTemplate(
+            mergedTemplate.template().settings(),
+            mergedTemplate.template().mappings(),
+            indicesService
         );
+
+        templateBuilder.settings(mergedSettingsBuilder);
+        return dataStream.copy().setSettings(mergedSettings).build();
```

Comment:
This seems to duplicate a lot of the task implementation starting at line 148. Anything we can do to reflector that to reduce the duplication?

Is it possible to push down the dry run logic all the way to the task execution so there's no diverging path?

Reply:
I'm on the fence about that. That's the way I had originally written it. In dry-run mode I would just return the original cluster state. But it felt odd running through the cluster state update mechanics just to effectively do template validation. So I pulled the template validation logic out and just ran it locally in dry-run mode. But I could see arguments either way.

Reply:
> This seems to duplicate a lot of the task implementation starting at line 148. Anything we can do to reflector that to reduce the duplication?

Which parts are you referring to? This code is all code I had moved out of that task logic around line 148 so that I could call it from both places w/o duplication.

Reply:
> Which parts are you referring to? This code is all code I had moved out of that task logic around line 148 so that I could call it from both places w/o duplication.

Sorry my bad -.- I cross referenced the class in my IDE but forgot I had checked out another branch. The looks good 👍🏻

Reply:
> Is it possible to push down the dry run logic all the way to the task execution so there's no diverging path?
> I'm on the fence about that. That's the way I had originally written it. In dry-run mode I would just return the original cluster state. But it felt odd running through the cluster state update mechanics just to effectively do template validation. So I pulled the template validation logic out and just ran it locally in dry-run mode. But I could see arguments either way.

That's fair enough. If you have tried it and it didn't look noticeably better then I'm happy to stick with it how it is unless anyone else has strong opinions :-)

Reply:
> If you have tried it and it didn't look noticeably better then I'm happy to stick with it how it is unless anyone else has strong opinions :-)

I tried it again just to be sure. The problem is that when the `updateSettingsExecutor` is executed, the listener on my `UpdateSettingsTask` is notified with the updated cluster state. And since I don't want to _actually_ update the cluster state in dry_run mode, that's the old cluster state. So I have no access to the data stream. I could do something like modify the UpdateSettingsTask with the new data stream so that it is available to the listener, but that gets kind of ugly (and those tasks aren't meant to be updated).